### PR TITLE
Update URLPopover role and focus return

### PR DESCRIPTION
**Pull request by talldan**
## What?
Fixes a few a11y things for the social links block (that I noticed when working on it recently) and media placeholders

Social Links:
- The social links block uses a `URLPopover`, which didn't have a role (I think it should be role="dialog") and the opener button also didn't have `aria-haspopup` specified, so I've added it with the value 'dialog'.
- The URLPopover is now labelled. I'm open to feedback on the labelling.
- The URLPopover in the social links block doesn't return focus to the opener button when it's closed. This PR makes it do that.
- If you press backspace too many times in the url popover, it deletes the block, which I found unusual, and it feels like a bug, so I've removed that behavior. Input should be constrained in the popover. I think this is probably a hangover from a past implementation where the URLPopover used to automatically open.

Media Placeholder (e.g. on the image block):
- The 'Insert from URL' button had similar issues, it opens a `URLPopover` with no role, and there's no `aria-haspopup` on the opener button. This is fixed.
- There was also a focus loss when closing the popover using escape, now fixed. Submitting causes an image to load, so that's a little trickier, I might have to draw the line on fixing that one as this PR has already spiralled. 😄 
- The URLPopover is now labelled.

## Testing Instructions
(use a screenreader)

### Social links
1. Add a social links block
2. Navigate to the 'Add block' button in the social links block and select it
3. Add a WordPress block
4. Arrow down until the 'WordPress' button (I noticed that this could have better labelling). It should be announced as a 'dialog pop-up button' or similar. Select it
5. Focus should be transferred to what's now announced as a dialog
6. Press escape or type in a url and hit enter to submit
7. Focus should be transferred back to the button.
8. Reopen the dialog and press delete - focus should remain in the popover and the block isn't deleted

### Media placeholder
1. Insert an image block
2. Tab to the 'Insert from URL' button. It should be announced as a 'dialog pop-up button' or similar. Select it.
3. Focus should be transferred to what's now announced as a dialog
4. Press escape, focus should return back to the button

### CODE REVIEW

**Comment by Mamaduka** [comment]
> If you press backspace too many times in the url popover, it deletes the block, which I found unusual, and it feels like a bug, so I've removed that behavior. Input should be constrained in the popover. I think this is probably a hangover from a past implementation where the URLPopover used to automatically open.

The behavior was implemented in #50903.

**Comment by talldan** [comment]
> The behavior was implemented in https://github.com/WordPress/gutenberg/pull/50903.

Thanks for finding that, that's quite recent, so it surprises me.

I don't think there was any inconsistency though. If anything it creates more inconistency. Try the following:
1. Insert a buttons block
2. Open the link popover on a button
3. Press delete - note that the button block isn't removed

Same with other blocks that have link/url popovers.

Another inconsistency is that you can focus the button within the social link block (focusing the button inside the block is different to focusing the wrapper of the block), hit delete and nothing happens.

cc @richtabor @ntsekouras

**Comment by ntsekouras** [comment]
>I don't think there was any inconsistency though. If anything it creates more inconistency. Try the following:

>1. Insert a buttons block
>2. Open the link popover on a button
>3. Press delete - note that the button block isn't removed

I think that's a bit different because in social icons we can only have links and therefore we have to open the link popover. I think it's more comparable to pressing backspace in an empty button, which deletes the button block.. I have no strong opinions on this one, but personally I feel it fine..


> Another inconsistency is that you can focus the button within the social link block (focusing the button inside the block is different to focusing the wrapper of the block), hit delete and nothing happens.

I agree that this is not ideal, but I think we should fix this, instead of removing the other behavior.

File: packages/block-editor/src/components/url-popover/index.js [hunk 1]
```diff
@@ -71,6 +71,8 @@ const URLPopover = forwardRef(
 		return (
 			<Popover
 				ref={ ref }
+				role="dialog"
```
Comment (ntsekouras) [inline packages/block-editor/src/components/url-popover/index.js:74]: Is this something we should consider adding in the Popover component? According to [mdn](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role) it seems probably okay, but I'm not sure if that was a conscious decision to leave the consumers handle it.

Additionally we have some Popover usages that fit the mdn description, but not 100% sure. Example is the in-between inserter.. 

--cc @mirka @afercia

Comment (alexstine) [reply packages/block-editor/src/components/url-popover/index.js:74]: That's what happens when we let components scope creep. Not blaming anyone, just a statement of reflection we've all got to live with now. I think given the situation we're in, might have to manually add it. :(

Comment (talldan) [reply packages/block-editor/src/components/url-popover/index.js:74]: Popover is used for a lot of things, so I'm not sure it should be added everywhere. For example, I don't think `role=dialog` is supposed to be added to dropdown menus, which already have the `menu` role.

Perhaps if there were more distinct components, like `DialogPopover`, beyond that I'm not sure there's an easy way to enforce it.

**Comment by talldan** [comment]
> I think that's a bit different because in social icons we can only have links and therefore we have to open the link popover. I think it's more comparable to pressing backspace in an empty button, which deletes the button block.. I have no strong opinions on this one, but personally I feel it fine..

I disagree, but that's ok. I think I'll remove it from this PR to unblock it and we can discuss it on an issue.

To go into more detail, I don't agree that it's the same as the button block, text content is different to a url. An input within a dialog is also very different to rich text content.

When the user inserts a button block, the caret is by default within the text content and in many blocks deleting all text content results in the removal of the block, so it's an expected behavior.

For URLs in both the button and social link block, the user has to carry out a specific action to open the link dialog, clicking on a button. At this point the user understands they're in a dialog, so they would expect it to behave just like any other dialog, with all their input constrained. In no other dialog does deleting all the text remove the block.

**Comment by alexstine** [comment]
I agree with @talldan here. Backspacing inside a dialog should not remove a block. I'm not even sure how you make a case for good UX on something like that...

**Comment by talldan** [comment]
I've split out that change into another PR, so hopefully it fully unblocks this one - https://github.com/WordPress/gutenberg/pull/61344.

**Comment by afercia** [comment]
A couple questions:

- To my understanding, the URLPopover component is planned to be replaced by the LinkControl, is that correct? Cc @richtabor 
- By adding a `role="dialog"` and _not hiding_ from assistive technology the content outside the URLPopover, this will basically become a _non-modal_ dialog. While tabbing is constrained within the URLPopover, screen reader users can exit the URLPopover by using any of the several screen reader shortcuts to navigate content e.g. to jump to headings and I think with NVDA and JAWS they can also exit the URLPopover by simply arrowing. Are we OK with this @alexstine ? It wouldn't be a 100% compliant implementation.

**Comment by alexstine** [comment]
@afercia It tests well for me.

```
<div class="components-popover block-editor-url-popover is-positioned is-alternate" data-wp-c16t="true" data-wp-component="Popover" role="dialog" aria-label="Edit social link" tabindex="-1" style="position: absolute; top: 0px; left: 0px; transform: translateX(294px) translateY(186px) translateZ(0px);">
```

Unless `aria-modal="false"` is set, it gets true by default so the content outside is indeed restricted.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-modal

Thanks.

**Review by alexstine — APPROVED** [review]
@talldan Giving this an approve. Do these blocks have any type of E2E tests currently? If so, might be worth adding some basic focus tests.

**Comment by afercia** [comment]
> Unless aria-modal="false" is set, it gets true by default so the content outside is indeed restricted.

@alexstine I'm not sure that's correct. Actually, the default value of [aria-modal](https://www.w3.org/TR/wai-aria-1.2/#aria-modal) is false. Of course, when aria-modal is entirely omitted, there's no value and that _should_ be treated as `false` by screen readers.

I'm guessing the fact a dialog content is restricted anyways by some screen readers when they find a `role="dialog"` with no explicit `aria-modal="true"` set may come from some internal implementation of the screen reader, which I would argue doesn't fully follow the specification.

By checking the Accessibility tab in the Chrome dev tools, it clearly shows the value of the modal property is `false` in this PR. Here's the values the browser exposes:

ARIA Attributes:
- role: dialog
- aria-label: Edit social link

Computer Properties
- Name: Edit social link
- Role: dialogf
- modal: false

Screenshot:

Screenshot 2024-05-23 at 10 56 35

As such, I still think we should make a decision on whether to add an explicit aria-modal attribute or not.
- If we want this popover to be 'modal', then it should have an `aria-modal="true'` attribute.
- Otherwise, it can be omitted but it would be best to set an explicit `aria-modal="false'` IMHO.

One more question: should this semantics be added also to other popovers to insert links e.g. the LinkControl component?

**Comment by alexstine** [comment]
@afercia Yeah, it's a fair point. It looks like NVDA anyway treats `role="dialog"` always as a modal dialog even if you set `aria-modal="false"`. I just can't imagine screen readers pay this too much attention though because having to manually set `aria-modal="true"` is not something I've seen around the web often. I guess according to the MDN link, it is indeed false by default but I misread it. Screen reader support varies though, have a look.

https://a11ysupport.io/tech/aria/aria-modal_attribute#age-of-results

I say this gets merged, doesn't look like this would hurt anything. Just not sure it really does anything at this point.

Thanks.

**Comment by afercia** [comment]
> Screen reader support varies though, have a look.

Yes, `aria-modal` isn't fully supported by VoiceOver and it was even more buggy when the Modal component was implemented a few years ago. At that time, aria-modal made the content itself of the modal dialog hidden from assistive technologies. See https://github.com/WordPress/gutenberg/pull/6261#discussion_r185296762

That is the reason why the Modal component does not use `aria-modal`. Instead, [it uses a small helper](https://github.com/WordPress/gutenberg/blob/bf3720532a6d0d4759d474faf8ff8e3eda11b305/packages/components/src/modal/aria-helper.ts) to hide all the rest of the page by the means of `aria-hidden="true"`.

Regardless my question is still: Do we want this Popover to be 'modal'?
In that case, using only `aria-modal="true"` wouldn't be sufficient.
Otherwise, if we want this URLPopover to be a non-modal dialog, it can ship as is.

I still think that making the link editing experience is important so if `role="dialog"` is added to this popover, it should be added to all the other similar UIs to edit links e.g. the LinkControl and others, if any.

**Comment by talldan** [comment]
Apologies for not following up on this one for a while. It's been a busy time. I have more time for code contributions now, so following up on some older PRs.

> Do we want this Popover to be 'modal'?

While tab navigation is constrained, I don't think other parts of the UI are made inert. It seems like both `aria-modal` `true` and `false` would both be wrong based on the description on MDN, as it seems to require those qualities to be both true or false (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-modal). 

I personally felt that an important aspect of the PR is for the user to be aware the opener button will send focus to a popover context with constrained focus and that there affordances like escape to close the dialog, as without it I imagine it can be quite confusing. There aren't too many implementations like this across the codebase fortunately, I think the LinkControl and the color controls in the inspector are two that I can see.

For a lot of these popovers, they seem to sit in a grey area where they're closer semantically to menus, but not quite 😬

**Comment by afercia** [comment]
> While tab navigation is constrained, I don't think other parts of the UI are made inert. It seems like both `aria-modal` `true` and `false` would both be wrong based on the description on MDN, as it seems to require those qualities to be both true or false

Not sure I follow. Disclaimer: It could be because of the 33 Celsius (91 Fahrenheit) temeprature right now in my city.

**Comment by talldan** [comment]
> Not sure I follow. Disclaimer: It could be because of the 33 Celsius (91 Fahrenheit) temeprature right now in my city.

😅 

In the MDN article it says "If a dialog is not modal — there is no inert background **and** focus isn't confined to the dialog — either include aria-modal="false" or omit the attribute altogether."

In the case of URLPopover, focus is confined but the background isn't inert (at least it didn't seem to be when I checked), so perhaps it doesn't meet the criteria for `aria-modal` being `false` or `true`. My interpretation is that popovers like `URLPopover` and `LinkControl` should be making the background inert the same way that modals do.

Perhaps it's bad wording in the MDN article, the actual spec doesn't seem to clarify exactly what `aria-modal="false"` means, only that `aria-modal="true"` requires an inert background and confined focus when focus moves automatically into the popover (https://w3c.github.io/aria/#aria-modal).

**Comment by afercia** [comment]
Thanks @talldan for clarifying. Yes that wording could be improved.
My point is that, as you also pointed out, these popovers are sort of 'mixed' behaviors as in: they do implement constrained tabbing but they aren't really 'modal'.
- Keyboard users who don't use a screen reader are actually 'constrained' when tabbing.
- Screen reader users are not, because screen readers provide many ways to navigate a page that is not tabbing.

So basically there's an inconsistent user experience for keyboard users and screen reader users. That is why I asked if we want to make these popovers be really 'modal' for everyone and provide a consistent experience.

**Comment by talldan** [comment]
I think we're on the same page, and I have the same question. I can do some research into what it would take to make them have modal like behavior. I also wonder now whether dropdown menus should also be modal too.

**Comment by alexstine** [comment]
@talldan @afercia Let's go ahead and add `aria-modal="true"`. There is no way with NVDA for example to navigate outside of these popovers when `role="dialog"` exists. Other screen readers may differ but even via special object type navigation, there is no way to access the content outside. I believe popover should contain the same ARIA support the modal component uses. It may be visually different but not different at all to assistive tech users.

> The aria-modal attribute is used to indicate that the presence of a "modal" element precludes usage of other content on the page. For example, when a modal dialog is displayed, it is expected that the user's interaction is limited to the contents of the dialog, until the modal dialog loses focus or is no longer displayed.
When a modal element is displayed, assistive technologies SHOULD navigate to the element unless focus has explicitly been set elsewhere. Assistive technologies MAY limit navigation to the modal element's contents. If focus moves to an element outside the modal element, assistive technologies SHOULD NOT limit navigation to the modal element.
When a modal element is displayed, authors MUST ensure the interface can be controlled using only descendants of the modal element. In other words, if a modal dialog has a close button, the button should be a descendant of the dialog. When a modal element is displayed, authors SHOULD mark all other contents as inert (such as "inert subtrees" in HTML) if the ability to do so exists in the host language.

Worth noting, inert is not required by this attribute in the 1.2 version.

> When a modal element is displayed, authors SHOULD mark all other contents as inert (such as "inert subtrees" in HTML) if the ability to do so exists in the host language.

I do not think the word "should" is equal to the word "required".

Docs: https://www.w3.org/TR/wai-aria-1.2/#aria-modal

EDIT: If we do decide to mark the content inert, let's simply copy how the modal component does this already.

Thanks.

**Comment by afercia** [comment]
> EDIT: If we do decide to mark the content inert, let's simply copy how the modal component does this already.

Ideally, it would be nice to go a little further.
This entire conversation surfaces the need to have a modal _behavior_ that can be attached to components, not only to the modal dialog. In conversations with @ciampo during the recent WordCamp Europe, I heard there are plans to make a new version of the modal dialog component that is a composite component. It would be worth considering to abstract the 'modal' logic that makes the rest of the content inert from the modal dialog component and make it availanle for use in other components. However, I think such an approach would ork only for components that are rendered a direct children of the body (e..g so called 'portals').

**Comment by ciampo** [comment]
Thank you for the ping!

We are indeed in the process of rewriting a bunch of components (using `ariakit` for the underlying implementation). After reworking `TabPanel` (and introducing its successor, `Tabs`), we've been focussing on `CustomSelectControl` and `DropdownMenu`. Next up, we'll probably be looking at `ComboboxControl` and potentially `Modal` (which will likely see the introduction of a new, better-named, successor component — `Dialog`).
The medium-term vision is to refactor all popover-based components to use `ariakit` under the hood, which should bring several advantages, including better handling of modality (`ariakit` already offers a prop to toggle modality of its popovers) and better z-layering (without the need for hardcoded `z-index`es). It will take time, because we need to make sure that the changes that we introduce are backwards (and forwards) compatible, but those are our intentions.

And bonus — here's an article that I find as a useful reference when discussing popovers, dialogs and modality: [https://hidde.blog/dialog-modal-popover-differences/](https://hidde.blog/dialog-modal-popover-differences/)

**Comment by talldan** [comment]
I've added `aria-modal="true"` to `URLPopover` and will merge the PR 👍
